tooling/SecretManager.VisualStudio/ProjectLocalSecretsManager.cs

```diff
@@ -0,0 +1,117 @@
+        }
+
+        public string SanitizeName(string name) => name;
+
+        public IReadOnlyCollection<char> GetInvalidCharactersFrom(string name) => Array.Empty<char>();
+
+        public async Task AddSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValid(name);
+            var store = await GetStoreAsync();
+            if (store.ContainsKey(name))
+            {
+                throw new ArgumentException("A secret with this name already exists.", nameof(name));
+            }
+
+            store.Set(name, value);
+            cancellationToken.ThrowIfCancellationRequested();
+            store.Save();
+        }
+
+        public async Task SetSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValid(name);
+
+            var store = await GetStoreAsync();
+
+            store.Set(name, value);
+            cancellationToken.ThrowIfCancellationRequested();
+            store.Save();
+        }
+
+        public async Task<string> GetSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValid(name);
+            var store = await GetStoreAsync();
+            return store[name];
+        }
+
+        public async Task<IReadOnlyCollection<string>> GetSecretNamesAsync(CancellationToken cancellationToken = default)
+        {
+            var store = await GetStoreAsync();
+            return store.ReadOnlyKeys;
+        }
+
+        public async Task<IReadOnlyDictionary<string, string>> GetSecretsAsync(CancellationToken cancellationToken = default)
+        {
+            return await GetStoreAsync();
+        }
+
+        public async Task<bool> RemoveSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValid(name);
+
+            var store = await GetStoreAsync();
+
+            var result = store.Remove(name);
+
+            cancellationToken.ThrowIfCancellationRequested();
+            store.Save();
+
+            return result;
+        }
+
+        private void EnsureKeyNameIsValid(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (name.Length == 0)
+            {
+                throw new ArgumentException(nameof(name));
+            }
+        }
+
+        private async Task<SecretsStore> GetStoreAsync()
+        {
+            var userSecretsId = await _propertiesProvider.GetCommonProperties().GetEvaluatedPropertyValueAsync(UserSecretsPropertyName);
+
+            if (string.IsNullOrEmpty(userSecretsId))
+            {
+                // TODO how should this be handled?
+                throw new InvalidOperationException("UserSecrets cannot be read or altered on this project because 'UserSecretsId' is not set");
```

Comment:
I think the best thing to do would be to be as tolerant as possible.

If the project doesn't have a secrets ID and it's a 'read' then obviously the answer is 'emtpy. If the project doesn't have a secrets id and it's a 'write' we should try to generate one.

Reply:
I decided to added a throw here for now. Initially I was generating a guid and calling IProjectProperties.SetPropertyValueAsync. I pulled it out for now to avoid race conditions. Also, we need to address how this surfaces to the user in the UI.

Reply:
> Also, we need to address how this surfaces to the user in the UI.

It depends on the gesture but generally you would not surface this in the UI. If tooling can trivially fix a problem, it should.